manager/state/store/memory.go

```diff
@@ -70,18 +71,47 @@ var (
 		Tables: map[string]*memdb.TableSchema{},
 	}
 	errUnknownStoreAction = errors.New("unknown store action")
+
+	// WedgeTimeout is the maximum amount of time the store lock may be
+	// held before declaring a suspected deadlock.
+	WedgeTimeout = 30 * time.Second
 )
 
 func register(os ObjectStoreConfig) {
 	objectStorers = append(objectStorers, os)
 	schema.Tables[os.Table.Name] = os.Table
 }
 
+// timedMutex wraps a sync.Mutex, and keeps track of how long it has been
+// locked.
+type timedMutex struct {
+	sync.Mutex
+	lockedAt atomic.Value
+}
+
+func (m *timedMutex) Lock() {
+	m.lockedAt.Store(time.Now())
```

Comment:
Should this be `defer m.lockedAt.Store(time.Now())`?  Otherwise, if the lock is wedged (e.g. `m.Mutex.Lock()`) will block, then so long as something else attempts to call `timedMutex.Lock()`, `lockedAt` value will be reset and we won't necessarily be able to tell if something's wedged?